packages/grpc-js/src/server-call.ts

```diff
@@ -553,102 +553,100 @@ export class Http2ServerCallStream<
-      stream.removeListener('end', onEnd);
-      stream.removeListener('error', onEnd);
+        if (err !== undefined) {
+          reject({ code: Status.INTERNAL, details: err.message });
+          return;
+        }
 
-      if (err !== undefined) {
-        next({ code: Status.INTERNAL, details: err.message });
-        return;
-      }
+        if (receivedLength === 0) {
+          reject({
+            code: Status.INTERNAL,
+            details: 'received empty unary message',
+          });
+          return;
+        }
 
-      if (receivedLength === 0) {
-        next({
-          code: Status.INTERNAL,
-          details: 'received empty unary message',
-        });
-        return;
-      }
+        call.emit('receiveMessage');
 
-      call.emit('receiveMessage');
+        const requestBytes = Buffer.concat(body, receivedLength);
+        const compressed = requestBytes.readUInt8(0) === 1;
+        const compressedMessageEncoding = compressed ? encoding : 'identity';
+        const decompressedMessage = call.getDecompressedMessage(
+          requestBytes,
+          compressedMessageEncoding
+        );
 
-      const requestBytes = Buffer.concat(body, receivedLength);
-      const compressed = requestBytes.readUInt8(0) === 1;
-      const compressedMessageEncoding = compressed ? encoding : 'identity';
-      const decompressedMessage = call.getDecompressedMessage(
-        requestBytes,
-        compressedMessageEncoding
-      );
+        if (Buffer.isBuffer(decompressedMessage)) {
+          call.safeDeserializeMessage(decompressedMessage, resolve, reject);
+          return;
+        }
 
-      if (Buffer.isBuffer(decompressedMessage)) {
-        call.safeDeserializeMessage(decompressedMessage, next);
-        return;
+        decompressedMessage.then(
+          decompressed =>
+            call.safeDeserializeMessage(decompressed, resolve, reject),
+          (err: any) =>
+            reject(
+              err.code
+                ? err
+                : {
+                    code: Status.INTERNAL,
+                    details: `Received "grpc-encoding" header "${encoding}" but ${encoding} decompression failed`,
+                  }
+            )
+        );
       }
-
-      decompressedMessage.then(
-        decompressed => call.safeDeserializeMessage(decompressed, next),
-        (err: any) =>
-          next(
-            err.code
-              ? err
-              : {
-                  code: Status.INTERNAL,
-                  details: `Received "grpc-encoding" header "${encoding}" but ${encoding} decompression failed`,
-                }
-          )
-      );
-    }
+    });
   }
 
   private safeDeserializeMessage(
     buffer: Buffer,
-    next: (
-      err: Partial<ServerStatusResponse> | null,
-      request?: RequestType
-    ) => void
+    resolve: (
+      value: void | RequestType | PromiseLike<void | RequestType>
+    ) => void,
+    reject: (reason: any) => void
```

Comment:
I don't like passing `resolve` and `reject` in here. It's not a normal async flow and it makes the code logic more complicated. It would be better for this function to return a promise itself. In fact, I think that if you make this function `async` and replace the entire function body with `return this.deserializeMessage(buffer)`, any error will bubble up as a promise rejection, which can be handled by the caller. Alternatively, you could keep the `try...catch`, but `throw` the object that is currently passed to `reject`.

Reply:
This was in an attempt to make this function behave is the same way as before. I can see how this definitely makes it less understandable. I opted to the second option, keeping the `try...catch`, because I want to keep the Error mapping, and otherwise I'd have to do it twice